sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/PTransform.java

```diff
@@ -319,4 +319,16 @@ public OutputT expand(InputT input) {
       }
     };
   }
+
+  /** Like {@link #compose(SerializableFunction)}, but with a custom name. */
+  @Experimental
+  public static <InputT extends PInput, OutputT extends POutput>
```

Comment:
Your example is spot-on and represents probably the majority use case of `PTransform.compose`. I agree with you that in cases where you're defining a composed transform inline as part of a pipeline, passing the name to `PCollection.apply` is sufficient.

The specific case where I've found the lack of name a nuisance is when I want to return a transform from a method. Consider the following enum, which we use as a pipeline option:

```
public enum InputType {

  pubsub {

    /** Return a PTransform that reads from a Pubsub subscription. */
    public PTransform<PBegin, ResultWithErrors<PCollection<PubsubMessage>>> read(
        SinkOptions.Parsed options) {
      return PTransform.compose(input -> input
          .apply(PubsubIO.readMessagesWithAttributes().fromSubscription(options.getInput()))
          .apply(MapElementsWithErrors.ToPubsubMessageFrom.identity()));
    }
  },

  file {

    /** Return a PTransform that reads from local or remote files. */
    public PTransform<PBegin, ResultWithErrors<PCollection<PubsubMessage>>> read(
        SinkOptions.Parsed options) {
      return PTransform.compose(input -> input.apply(TextIO.read().from(options.getInput()))
          .apply(options.getInputFileFormat().decode()));
    }
  };

  public abstract PTransform<PBegin, ResultWithErrors<PCollection<PubsubMessage>>> read(
      SinkOptions.Parsed options);
}
```

My main class then calls:

```
input.apply(options.getInputType().read())
```

I'd like the name of the transform to indicate what the input type was rather than picking a static name that I pass into `apply`.

Reply:
The above example is perhaps too complex and specific to justify adding a new signature to the API, so let me take a step back.

This signature would be useful in any case where you want to return a simple transform from a method. A user might want to make a simple library of transforms that can be reused in different pipelines. In such a case, it's convenient to be able to give the transform you're returning a custom name.

So, a simpler example:

```
public class MyTransforms extends PTransform<...> {
  public static PTransform<...> transform1() {
    return PTransform.compose("MyTransforms.transform1", ...)
  }
}
```

It seems better to repeat the name once where we define the composite transform rather than having to repeat the name every time we use it in a pipeline:

```
input1.apply(MyTransforms.transform1())...

input2.apply(MyTransforms.transform1())...
```

It's a better experience for the end pipeline developer if the composite transform has a name rather than displaying as "AnonymousTransform", so seems worth it to me to add an additional signature here to encourage the practice. Perhaps the Javadoc should provide an example like the above and discuss that the name specified here is a fallback in case `apply` is called without an explicit name.

If, after the above discussion, you still feel iffy about it, I'm perfectly happy to go ahead and close this. I appreciate your review and thoughts.

Reply:
First off-- thanks for taking the time to write up the motivation. I believe I have a good understanding of the scenario now.

So, have two ways of specifying a composite transform:

1. By defining a class which extends `PTransform<TIn, TOut>` and filling in the `expand()` method.
2. Using this `PTranform.compose()` helper syntax.

Defining a class gives access to all features of the PTransform spec (naming, validation, display data, multiple inputs), while the helper syntax offers a very concise syntax for the simplest cases.

My general thoughts:

a. We shouldn't add verbosity to the shorthand syntax as it removes the utility of it.
b. We shouldn't add a bunch of complexity to the shorthand but and instead defer to the full syntax.

In this case, having an overload with a String name is still a useful shorthand and doesn't violate either of the above. And you've defined a real scenario where it provides value. So: I'm convinced.